src/platform/intel/cavs/lib/clk.c

```diff
@@ -145,24 +129,67 @@ static inline void set_cpu_current_freq_idx(int freq_idx, bool release_unused)
 {
 	select_cpu_clock(freq_idx, true);
 }
+#endif /* !(CONFIG_CAVS_LPRO_ONLY) */
 
 static void platform_clock_low_power_mode(int clock, bool enable)
 {
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+	int current_freq_idx = get_current_freq_idx(clock);
+	int freq_idx = *cache_to_uncache(&active_freq_idx);
+
+	if (enable && current_freq_idx != CPU_LPRO_FREQ_IDX)
+		/* LPRO requests are fast, but requests for other ROs
+		 * can take a lot of time. That's why it's better to
+		 * not release active clock just for waiti,
+		 * so they can be switched without delay on wake up.
+		 */
+		select_cpu_clock(CPU_LPRO_FREQ_IDX, false);
+	else if (!enable && current_freq_idx != freq_idx)
+		select_cpu_clock(freq_idx, false);
+#endif /* CONFIG_CAVS_USE_LPRO_IN_WAITI */
 }
-#endif
 
-void platform_clock_on_waiti(void)
+static int clock_platform_set_cpu_freq(int clock, int freq_idx)
 {
+	set_cpu_current_freq_idx(freq_idx, true);
+	return 0;
 }
 
-void platform_clock_on_wakeup(void)
+void platform_clock_on_waiti(void)
 {
+/* do nothing for LPRO_ONLY mode */
+#if !(CONFIG_CAVS_LPRO_ONLY)
+	/* perform the possible dynamic clock switching */
+	if (*cache_to_uncache(&active_freq_idx) != CPU_LPRO_FREQ_IDX &&
+	    !pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID))
+		set_cpu_current_freq_idx(CPU_LPRO_FREQ_IDX, true);
+	else if (*cache_to_uncache(&active_freq_idx) != CPU_HPRO_FREQ_IDX &&
+		 pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID))
+		set_cpu_current_freq_idx(CPU_HPRO_FREQ_IDX, true);
+
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+	/* check if waiti HPRO->LPRO switching is needed */
+	pm_runtime_put(CORE_HP_CLK, cpu_get_id());
+#endif /* CONFIG_CAVS_USE_LPRO_IN_WAITI */
+#endif /* !(CONFIG_CAVS_LPRO_ONLY) */
 }
 
-static int clock_platform_set_cpu_freq(int clock, int freq_idx)
+void platform_clock_on_wakeup(void)
 {
-	set_cpu_current_freq_idx(freq_idx, true);
-	return 0;
+/* do nothing for LPRO_ONLY mode */
+#if !(CONFIG_CAVS_LPRO_ONLY)
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+	/* check if LPRO->HPRO switching back is needed */
+	pm_runtime_get(CORE_HP_CLK, cpu_get_id());
+#else
+	int current_idx = get_current_freq_idx(CLK_CPU(cpu_get_id()));
+	int target_idx = *cache_to_uncache(&active_freq_idx);
+
+	/* if no LPRO_IN_WAITI, restore the active cpu freq_idx manually */
+	if (current_idx != target_idx)
+		set_cpu_current_freq_idx(target_idx, true);
+#endif /* CONFIG_CAVS_USE_LPRO_IN_WAITI */
+#endif /* !(CONFIG_CAVS_LPRO_ONLY) */
```

Comment:
It looks like most of these function bodies are conditionally compiled code, this may be better if you had two version of each function here, one for static clock (LPRO or HP clock) the other for dynamic clock.

Reply:
We don't have HPRO only mode, so basically
1) by default clock switching is used, at least we need to switch to LPRO when in D0i3, and switch back to HPRO when performing D0i3 -> D0.
2) if USE_LPRO_IN_WAITI is selected, it will switch to LPRO during WFI, no matter in D0 or in D0i3.
3) LPRO only for debug purpose only, where only LPRO is used and therefore no clock switching is needed for all scenarios including 1) and 2).

The preserving of the LPRO_ONLY mode increases the complexity of this conditionally compiling, that's one of the reason why I wanted to remove the debug mode initially, @mmaka1 @lbetlej

Reply:
We want to keep LPRO_ONLY mode. It would improve readability if there are three ver. of the function compiled conditionally as @lgirdwood noted.

Reply:
> We want to keep LPRO_ONLY mode. It would improve readability if there are three ver. of the function compiled conditionally as @lgirdwood noted.

OK, let me change to use 3 separated versions.